### PR TITLE
feat(modal): dismissAll

### PR DESCRIPTION
**Pull request by divdavem**
This PR implements the feature requested in #1963

Before submitting a pull request, please make sure you have at least performed the following:

 - [x] read and followed the [CONTRIBUTING.md](https://github.com/ng-bootstrap/ng-bootstrap/blob/master/CONTRIBUTING.md) guide.
 - [x] built and tested the changes locally.
 - [x] added/updated any applicable tests.
 - [x] added/updated any applicable API documentation.
 - [x] added/updated any applicable demos.

### CODE REVIEW

File: src/modal/modal-stack.ts [hunk 1]
```diff
@@ -22,6 +22,7 @@ import {NgbActiveModal, NgbModalRef} from './modal-ref';
 export class NgbModalStack {
   private _windowAttributes = ['ariaLabelledBy', 'backdrop', 'centered', 'keyboard', 'size', 'windowClass'];
   private _backdropAttributes = ['backdropClass'];
+  public modalRefs: NgbModalRef[] = [];
```
Comment (pkozlowski-opensource) [inline src/modal/modal-stack.ts]: Why is this public? Is it for tests only? If so please check my comment in the test file.

File: demo/src/app/components/modal/modal.component.ts [hunk 2]
```diff
@@ -16,6 +16,9 @@ import {DEMO_SNIPPETS} from './demos';
       <ngbd-example-box demoTitle="Components as content" [snippets]="snippets" component="modal" demo="component">
           <ngbd-modal-component></ngbd-modal-component>
       </ngbd-example-box>
+      <ngbd-example-box demoTitle="Dismiss all modal windows" [snippets]="snippets" component="modal" demo="dismissall">
+          <ngbd-modal-dismissall></ngbd-modal-dismissall>
+      </ngbd-example-box>
```
Comment (pkozlowski-opensource) [inline demo/src/app/components/modal/modal.component.ts]: I'm not sure if we need a dedicated demo for this new feature - I would rather drop it as IMO it doesn't add enough value.

File: src/modal/modal.spec.ts [hunk 3]
```diff
@@ -248,6 +249,38 @@ describe('ngb-modal', () => {
         expect(fixture.nativeElement).not.toHaveModal();
       });
 
+      it('should dismiss with dismissAll', fakeAsync(() => {
```
Comment (pkozlowski-opensource) [inline src/modal/modal.spec.ts]: This tests go through low-level APIs `NgbModalStack` and asserts on "private" (AFAIK) API (`modalRefs`).

I would prefer this test to be written as all the existing tests and go through  `NgbModal` - just open one modal, call `dismissAll(...)` on `NgbModal` and assert that we don't have any modals left in the DOM. We don't need much more for now, until we support stacked modals.

**Comment by pkozlowski-opensource** [comment]
@divdavem thnx for the PR - the implementation looks good, I would just:
* drop the demo;
* re-work test to expose less of APIs as public.

More info in the in-line comments.

File: src/modal/modal-stack.ts [hunk 4]
```diff
@@ -57,6 +59,19 @@ export class NgbModalStack {
...
+      const index = this.modalRefs.indexOf(ngbModalRef);
+      if (index > -1) {
+        this.modalRefs.splice(index, 1);
+      }
+    };
+    this.modalRefs.push(ngbModalRef);
+    ngbModalRef.result.then(unregisterOpenedModalRef, unregisterOpenedModalRef);
+  }
+
+  public dismissAll(reason?: any) { this.modalRefs.forEach(ngbModalRef => ngbModalRef.dismiss(reason)); }
```
Comment (pkozlowski-opensource) [inline src/modal/modal-stack.ts]: nit: drop `public` - everything is public by default.

**Comment by divdavem** [comment]
@pkozlowski-opensource Thank you for your review! I have done the changes you requested.

**Pull request closed without merging**